[Order form] Use order item data for products in order creation/editing

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11455
Investigation: pdfdoF-4rm-p2
<!-- Id number of the GitHub issue this PR addresses. -->

## Description

This PR addresses an issue where iOS uses `Product` price, instead of the `Order Item ` price in the Order Creation / Editing view.

It results in a couple of issues when **Yes, I will enter prices inclusive of tax** setting is enabled.

![image](https://github.com/user-attachments/assets/4608e74a-88b0-4fdc-b5af-6cdac2cace6e)

1. iOS Order Creation/Editing view is inconsistent with Android, and Web. 
2. iOS Order Creation/Editing view is inconsistent with the iOS Order Details view.
3. Product prices displayed in the Products Section include taxes while product prices displayed in Order Summary exclude taxes, which results in different Product totals shown.
    3.1 When discounts are applied, discounted totals in the Product Section are calculated with included taxes, again resulting in different product prices displayed in Order Summary.
    3.2. Individually-priced bundle products are shown with full price, rather than an individual price within the bundle, which sums up to a different total that is shown in the Order Summary.

### Solution

The simplest solution I found is to pass `a single pre-tax pre-discount order item price` to `CollapsibleProductRowCardViewModel` rather than `Product.price`. This mimics what Android implemented (https://github.com/woocommerce/woocommerce-android/blob/1e2829c413233e7a1a8602137ea8b2858b377386/WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt#L126) with introducing `basePrice` to `Order Item`. This is also could be considered a continuation of another fix https://github.com/woocommerce/woocommerce-ios/pull/11012 which started using this single order item subtotal price to address order total calculation issues.

`basePrice` behaves as we would expect because it:
- Shows a consistent price independent of **Yes, I will enter prices inclusive of tax**
- Does not change when the quantity increases
- Does not contain a discount
- Works for simple products, variable products, individually priced product bundle products
- Correctly sums up to what we see in the Order Summary at the bottom of the Order view
- Reaches parity with Android, and web

Other existing `Order Item` properties do not satisfy these requirements:

- OrderItem subtotal - excludes discounts and depends on item quantity
- OrderItem total - includes discounts and depends on item quantity
- OrderItem price - includes discounts
- Product price - always static and doesn't take into account whether taxes are included or excluded from the product

Other not-addressed issues:

- [x] https://github.com/woocommerce/woocommerce-ios/pull/11431 also started to implement loading states for Product and Order prices when the Order is syncing. There's also a related issue where the order total is momentarily wrong while Order is syncing (for example when quantity is changed) I'm leaving it for another task & PR: https://github.com/woocommerce/woocommerce-ios/pull/13340
- [x] I found the issue on Android and iOS where two order items for $15 total to $30.01 while on the web it totals correctly to $30.00. There may be some issues in how taxes are calculated. I will check if such an issue is reported, if not, I'll report it. (**Edit:** Looks like this issue is known https://github.com/woocommerce/woocommerce-ios/pull/11012#issuecomment-1780280078 peaMlT-ec-p2)
- [x] I found the issue on Android and iOS where Coupon Discount is applied differently compared to the web, resulting in completely different Order Totals.  I will check if such an issue is reported, if not, I'll report it. (**Edit:** peaMlT-KO-p2)
- [x] https://github.com/woocommerce/woocommerce-ios/issues/11393 addressed in https://github.com/woocommerce/woocommerce-ios/pull/13313

### Simple Product

The product price is correctly displayed before tax which matches the "Products" line in the Order Summary

| Before | After |
|--------|-------|
| ![Simple Product - Before](https://github.com/user-attachments/assets/e2be0716-fadf-4208-b10d-72c318bc13ac) | ![Simple Product - After](https://github.com/user-attachments/assets/e9b3737f-473a-4c9c-8465-893b34958994) |

### Simple Product + Discount

The product price is correctly displayed before which matches the "Products" line in the Order Summary

| Before | After |
|--------|-------|
| ![Simple Product - Product Discount - Before](https://github.com/user-attachments/assets/fdefe49e-50f7-4e20-9729-8956fd67dec2) | ![Simple Product - Product Discount - After](https://github.com/user-attachments/assets/616da414-6ff0-45be-acd4-cd052eb9e8d6) |

### Variable Product + Discount

| Before | After |
|--------|-------|
| ![Variable Product - Product Discount - Before](https://github.com/user-attachments/assets/5fc5208f-6675-4b5d-b08e-de0f7e11b0e4) | ![Variable Product - Product Discount - After](https://github.com/user-attachments/assets/b69fccc0-4e05-4823-a871-a49804516915) |

### Product Bundle + Child item custom price

- **Before**: Bundle products appear for $100, child items for $55 and $16 which totals to $171, however "Products" line shows $152.55
- **After**: Bundle product appears for $90.91, child items for $50 and $11.64, which correctly totals to $152.55 as in "Products" line 

| Before | After |
|--------|-------|
| ![Bundled Product - Before](https://github.com/user-attachments/assets/f3423920-2802-4470-a925-47813d318b88) | ![Bundled Product - After](https://github.com/user-attachments/assets/7d09ae1d-14b6-4787-8bdb-e4822f35d723) |

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Checklist borrowed from pdfdoF-4rm-p2:

1. With the Prices entered with tax setting – **Yes, I will enter prices inclusive of tax**
    1.1 When adding an item to an order –
        - The item's total should be equal to the item's price, excluding taxes.
        - Taxes total should be calculated based on the item's price excluding taxes.
    1.2 When incrementing a quantity of the item –
        - The items' total should be updated accordingly, excluding taxes.
        - Taxes total should be calculated based on the sum of items' prices excluding taxes.
    1.3 When adding a discount to the item –
        - The item's total should be reduced according to the discount amount.
        - Taxes total for that item should be calculated based on the item's total excluding taxes reduced by the discount amount.

2. With the Prices entered with tax setting – **No, I will enter prices exclusive of tax**
    2.1 When adding an item to an order –
         - The item's total should be equal to the item's price, excluding taxes.
         - Taxes total should be calculated based on the item's price excluding taxes.
    2.2 When incrementing a quantity of the item –
        - The items' total should be updated accordingly, excluding taxes.
        - Taxes total should be calculated based on the sum of items' prices excluding taxes.
    2.3 When adding a discount to the item –
        - The item's total should be reduced according to the discount amount.
        - Taxes total for that item should be calculated based on the item's total excluding taxes reduced by the discount amount.

**Testing steps** (Borrowed from https://github.com/woocommerce/woocommerce-ios/pull/11431)

- Enable **Yes, I will enter prices inclusive of tax** tax setting
- Go to the Orders tab
- Tap on `+` to create an order
- Add at least one product to the order -> Confirm the product appears in the order form with the expected product details, and with price excluding tax
- Expand any product card and confirm you can update the product as expected (change quantity, add discount, remove product)
- Tap "Create" to create the order -> Confirm the created order contains the products and quantities you added, and make a note of the prices shown in order details
- Tap "Edit" to edit the order -> Confirm the order loads with the expected products and quantities, and with prices matching what was shown in order details

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Add pricePreDiscount to OrderItem to display expected single line item price

When making order, we need to show a order item price for a single item without discounts and taxes. There's no property to support all of these requirements so we need to calculate it ourselves. This mimics the functionality that Android already has.

staskus committed:
Use pre-discount order item price when displaying products in Order editing/creation flow

When we have Tax setting "Yes, I will enter prices inclusive of tax" enabled, using product.price in Order view results in different information shown in Product List and Order Summary within Order View. Order Summary shows product prices without taxes, while Product List shows full price including the tax.

Use pre-discount single order item price which mimics the behavior on Android and on the web when displaying products within order.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Update EditableOrderViewModelTests.swift

staskus commented:
@rachelmcr I would appreciate it if you could take a look at this 🙇 . 

I opted to make as simple changes as possible to address the main issue - showing misleading prices in the product section when **Yes, I will enter prices inclusive of tax** setting is enabled. 

In your previous PR https://github.com/woocommerce/woocommerce-ios/pull/11431 you made further improvements around loading states and chose to pass additional properties to the collapsable card rather than using the existing `price` parameter. I will address loading states and other issues separately. However, as I tested it works fine passing single order item price to `price` parameter to make it all work. Let me know how you see it. Thanks again!

staskus committed:
Move OrderItem+PricePreDiscount to Yosemite and reuse for ProductInputTransformer

When calculating prices, Yosemite already uses pricePreDiscount logic within baseSubtotal method.

iamgabrielma commented on Yosemite/Yosemite/Model/Orders/OrderItem+PricePreDiscount.swift:
```diff
@@ -0,0 +1,26 @@
+    ///
+    /// - Parameter orderItem
+    ///
+    var pricePreDiscount: NSDecimalNumber {
```
The name of `pricePreDiscount` seems to imply that there's always a discount to be applied after, which may not be the case. Also makes no reference to being affected by tax. What about using a name like `netPrice`, `basePrice`, or similar?

staskus replied:
Good point. The only reason I named it this way was because Android used this naming https://github.com/woocommerce/woocommerce-android/blob/1e2829c413233e7a1a8602137ea8b2858b377386/WooCommerce/src/main/kotlin/com/woocommerce/android/model/Order.kt#L126

Android chose this naming because `orderItem.price` is a single item price that includes discount, `orderItem.pricePreDiscount` is same as `orderItem.price` minus discount.

iamgabrielma commented on Yosemite/Yosemite/Stores/Order/ProductInputTransformer.swift:
```diff
@@ -137,7 +137,7 @@ public struct ProductInputTransformer {
                                          quantity: quantity,
                                          discount: discount ?? defaultDiscount,
-                                         baseSubtotal: baseSubtotal(on: item),
+                                         baseSubtotal: item.pricePreDiscount.decimalValue,
```
While not part of this PR, the same as above: How about `netPrice` or `basePrice` for the property? I'd say `baseSubtotal` feels off since both base and subtotal should mean the same price-wise ( a base price without any modifiers).

iamgabrielma reviewed (approved):
Thanks for taking care of this! 

I found two things that behave oddly, but I think both precedes your changes and seem related to displaying the price without taking into account the store settings regarding inclusive/exclusive of taxes:

1. When testing custom amounts note how using `Yes, I will enter prices inclusive of tax` still allows us to add additional tax on top of a custom amount, since can be edited by toggling the button.

| Custom amount  | Toggle taxes |
|---|---|
| ![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-07-16 at 12 08 23](https://github.com/user-attachments/assets/195e2a99-aa82-4fc3-bbe2-d63b3c9b7816)  |  ![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-07-16 at 12 08 35](https://github.com/user-attachments/assets/8f0245e1-fe92-4603-bc78-8d8afecc40ed) |

2. Having products inclusive/exclusive of taxes seems to be ignored if we "Set new tax rate". In this example prices should already have taxes baked in, but upon setting a new tax it adds it to the current price 🤔 

 
![Simulator Screen Recording - iPad Pro (12 9-inch) (6th generation) - 2024-07-16 at 12 30 44](https://github.com/user-attachments/assets/73354ccf-ec24-4fe4-b5b6-53c8d93a7f62)

I don't have full context of the custom amounts/set tax features, but as mentioned these exist before your PR, so we should log/address them separately

staskus committed:
Update RELEASE-NOTES.txt

staskus commented:
Thanks for the review, @iamgabrielma!

> When testing custom amounts note how using `Yes, I will enter prices inclusive of tax` still allows us to add additional tax on top of a custom amount, since can be edited by toggling the button.

I think it's fine behavior, no? Regardless of this setting, we still allow users to select whether a tax will be applied to the Custom Amount or not. Even if `Yes, I will enter prices inclusive of tax` is enabled but we choose not to charge taxes, then the taxes are not charged for this Custom Amount.

Your expected behavior would be that if `Yes, I will enter prices inclusive of tax` enabled, then tax should be applied by default to Custom Amount? 

> Having products inclusive/exclusive of taxes seems to be ignored if we "Set new tax rate". In this example prices should already have taxes baked in, but upon setting a new tax it adds it to the current price 🤔

I see what you mean and that's a very good point. I agree that if the product already includes taxes, then the total shouldn't change, just the `pricePreDiscount`. I cannot get it to work in the same way on wp-admin as well. It may be a problematic area in the app.

I created an issue: https://github.com/woocommerce/woocommerce-ios/issues/13342 and may file a CUW p1721127325826489-slack-C025A8VV728

**Edit:** This issue is also reported on the web (https://github.com/woocommerce/woocommerce/issues/30702). Even if it's a bug, it's not an app issue.

staskus committed:
Rename pricePreDiscount to basePrice

staskus committed:
Merge branch 'trunk' into fix/11389-use-order-item-data-in-order-form

rachelmcr commented:
I haven't looked closely at the code to see what's causing it, but these changes aren't working well for me in testing. Here's an example:

* On my test store, my product prices include tax and tax is based on my store address.
* My store address is in a location with a 6% sales tax.
* I have a product that is $15.

When I add that product to an order on the web or in `trunk`, the order total is $15 ($14.15 for the product + $0.85 in tax). However, when I add that product to an order in this branch, it still shows the product with the price including tax, and it adds more tax on top of it for an order total of $15.90 ($15 for the product + $0.90 in tax).

Web|App (`trunk`)|App (this branch)
-|-|-
![image](https://github.com/user-attachments/assets/06f8a9fa-6807-4e4c-9ec7-0cb6783adb25)|![image](https://github.com/user-attachments/assets/d8469036-84a1-413f-829b-5416181860ea)|![Simulator Screenshot - iPhone 15 Plus - 2024-07-16 at 18 32 34](https://github.com/user-attachments/assets/9bbf52a3-d488-484c-8935-5f39ad6d7f86)

I'm out of time to review this further today, but I'll take another look tomorrow to see if I can sort out why it isn't working for me.

rachelmcr commented:
Nevermind, you can ignore my last comment. I'm not sure what was happening with my testing on this branch earlier, but it's now all working as expected. Looks good!

The only issue I've seen is with the "Price after discount" amount when editing a product discount where the product quantity is > 1, but it looks like that is resolved with https://github.com/woocommerce/woocommerce-ios/pull/13313.

> In your previous PR https://github.com/woocommerce/woocommerce-ios/pull/11431 you made further improvements around loading states and chose to pass additional properties to the collapsable card rather than using the existing price parameter. I will address loading states and other issues separately. However, as I tested it works fine passing single order item price to price parameter to make it all work. Let me know how you see it.

Your solution looks good to me! I can't remember why I approached it the way I did before, and this approach looks clean and easy to follow.

staskus commented:
Thanks for the review @rachelmcr! Appreciate that 🙇 

>  I'm not sure what was happening with my testing on this branch earlier, but it's now all working as expected. Looks good!

Good to hear! Not sure what happened there as well 🤔 

> The only issue I've seen is with the "Price after discount" amount when editing a product discount where the product quantity is > 1, but it looks like that is resolved with https://github.com/woocommerce/woocommerce-ios/pull/13313.

Yes, it was resolved there. I haven't merged this branch with trunk after merging the other PR.

staskus committed:
Merge branch 'trunk' into fix/11389-use-order-item-data-in-order-form
